### PR TITLE
Add cpp data codec

### PR DESCRIPTION
**Pull request by xige-16**
issue: #19092 

1. add insert/index data codec in core/storage
2. add cpp chunk manager to sync remote and disk data after codec data

Signed-off-by: xige-16 <xi.ge@zilliz.com>
Co-Authored-By: zhagnlu lu.zhang@zilliz.com

### CODE REVIEW

File: internal/core/src/storage/FieldData.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    *str_size = length;
+}
+
+std::unique_ptr<Payload>
+FieldData::get_payload() const {
+    AssertInfo(array_ != nullptr, "null arrow array");
+    auto raw_data_info = std::make_unique<Payload>();
+    raw_data_info->rows = array_->length();
+    raw_data_info->data_type = data_type_;
+    std::cout << "length = " << array_->length() << std::endl;
```
Comment (jiaoew1991) [inline internal/core/src/storage/FieldData.cpp:78]: use `easylogging` other than stdout?

File: internal/core/src/storage/PayloadWriter.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+void
+PayloadWriter::add_payload(const Payload& raw_data) {
+    AssertInfo(output_ == nullptr, "payload writer has been finished");
+    AssertInfo(column_type_ == raw_data.data_type, "mismatch data type");
+
+    // lazy init for float/binary vector
+    // Compatible with codec logic implemented by go
+    if (milvus::datatype_is_vector(column_type_) && !dimension_.has_value()) {
+        AssertInfo(raw_data.dimension.has_value(), "empty dimension");
+        init_dimension(raw_data.dimension.value());
```
Comment (jiaoew1991) [inline internal/core/src/storage/PayloadWriter.cpp:66]: should we add some checker to find errors early, like checking the size of raw data is right

File: internal/core/src/storage/Event.cpp [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    AssertInfo(ret.ok(), "read input stream failed");
+    int data_size = milvus::datatype_sizeof(data_type) * row_num;
+    auto insert_data_bytes = input->Read(data_size);
+    auto insert_data = reinterpret_cast<const uint8_t*>(insert_data_bytes.ValueOrDie()->data());
+    std::shared_ptr<arrow::ArrayBuilder> builder = nullptr;
+    if (milvus::datatype_is_vector(data_type)) {
+        builder = CreateArrowBuilder(data_type, dimension);
+    } else {
+        builder = CreateArrowBuilder(data_type);
+    }
```
Comment (jiaoew1991) [inline internal/core/src/storage/Event.cpp:289]: extract it as a util function maybe better 🤔

File: internal/core/src/common/Consts.h [hunk 4]
```diff
@@ -25,3 +25,9 @@ const milvus::PkType INVALID_PK;  // of std::monostate if not set.
 // TODO: default field start id, could get from config.yaml
 const int64_t START_USER_FIELDID = 100;
 const char MAX_LENGTH[] = "max_length";
+const char ORIGIN_SIZE_KEY[] = "original_size";
+const char INDEX_BUILD_ID_KEY[] = "indexBuildID";
+const char INDEX_VERSION_KEY[] = "version";
+const char INDEX_NAME_KEY[] = "indexName";
+const char INDEX_FILE_KEY[] = "key";
+const char INDEX_ID_KEY[] = "indexID";
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/common/Consts.h]: > format to be uniform

Unity will have compatibility issues

File: internal/core/src/storage/Event.cpp (hunk 3 above)
Comment (xige-16) [reply internal/core/src/storage/Event.cpp:289]: > extract it as a util function maybe better thinking

？

File: internal/core/src/storage/DataCodec.h [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+enum CodecType {
+    InvalidCodecType = 0,
+    InsertDataType = 1,
+    IndexDataType = 2,
+};
+
+class DataCodec {
+ public:
+    explicit DataCodec(std::shared_ptr<FieldData> data, CodecType type) : field_data_(data), codec_type_(type) {
```
Comment (cydrain) [inline internal/core/src/storage/DataCodec.h:45]: you can define "FieldDataPtr"

File: internal/core/src/storage/DataCodec.h [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    virtual ~DataCodec() = default;
+
+    // Serialized data can be written directly to remote or local disk
+    virtual const std::vector<uint8_t>
```
Comment (cydrain) [inline internal/core/src/storage/DataCodec.h:51]: should not use "std::vector<uint8_t>", use "std::unique_ptr<uint8_t[]>"

File: internal/core/src/storage/Event.h [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    milvus::Timestamp timestamp_;
+    EventType event_type_;
+    int32_t event_length_;
+    int32_t next_position_;
+
+    EventHeader() {
+    }
+    explicit EventHeader(PayloadInputStream* input);
+
+    std::vector<uint8_t>
```
Comment (cydrain) [inline internal/core/src/storage/Event.h:41]: use "std::unique_ptr<uint8_t[]>"

File: internal/core/src/storage/Event.h [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    Serialize();
+};
+
+struct DescriptorEventData {
+    DescriptorEventDataFixPart fix_part;
+    int32_t extra_length;
+    std::vector<uint8_t> extra_bytes;
+    std::map<std::string, std::string> extras;
```
Comment (cydrain) [inline internal/core/src/storage/Event.h:66]: if need not keep data order, should use std::unordered_map"

File: internal/core/src/storage/Event.h [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+    DescriptorEventData() {
+    }
+    explicit DescriptorEventData(PayloadInputStream* input);
+
+    std::vector<uint8_t>
+    Serialize();
+};
+
+struct InsertEventData {
```
Comment (cydrain) [inline internal/core/src/storage/Event.h:77]: "InsertEventData" and "IndexEventData" are redundant structures.
Only define "EventData" is fine, add "EventType" in it.

File: internal/core/src/storage/Event.cpp [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+           sizeof(data.fix_part.end_timestamp) + sizeof(data.fix_part.data_type);
+}
+int
+GetFixPartSize(InsertEventData& data) {
+    return sizeof(data.start_timestamp) + sizeof(data.end_timestamp);
+}
+
+int
+GetEventHeaderSize(EventHeader& header) {
+    return sizeof(header.event_type_) + sizeof(header.timestamp_) + sizeof(header.event_length_) +
```
Comment (cydrain) [inline internal/core/src/storage/Event.cpp:41]: why not use sizeof(header) ?

File: internal/core/src/storage/Event.cpp [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+    assert(ast.ok());
+    ast = input->Read(sizeof(event_type_), &event_type_);
+    assert(ast.ok());
+    ast = input->Read(sizeof(event_length_), &event_length_);
+    assert(ast.ok());
+    ast = input->Read(sizeof(next_position_), &next_position_);
+    assert(ast.ok());
+}
+
+std::vector<uint8_t>
```
Comment (cydrain) [inline internal/core/src/storage/Event.cpp:78]: suggest to use "std::unique_ptr<uint8_t[]"
if return "std::vector<uint8_t>", then should "return std::move(res)"

File: internal/core/src/storage/Event.h [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+    InsertEvent() {
+    }
+    explicit InsertEvent(PayloadInputStream* input, DataType data_type);
+
+    std::vector<uint8_t>
+    Serialize();
+};
+
+using IndexEvent = InsertEvent;
```
Comment (cydrain) [inline internal/core/src/storage/Event.h:114]: it's weird to let all event types equal to InsertEvent.
suggestion 1:
```
# change InsertEvent to BaseEvent
using IndexEvent = BaseEvent
using DeleteEvent = BaseEvent
...
```
suggestion 2:
```
# change InsertEvent to BaseEvent"
struct IndexEvent : public BaseEvent {
}
struct DeleteEvent : public BaseEvent {
}
```

I prefer suggestion 2

File: internal/core/src/storage/Event.cpp [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+    memcpy(res.data() + offset, &extra_length, sizeof(extra_length));
+    offset += sizeof(extra_length);
+    memcpy(res.data() + offset, extra_bytes.data(), extra_bytes.size());
+
+    return res;
+}
+
+InsertEventData::InsertEventData(PayloadInputStream* input, int event_length, DataType data_type) {
+    auto ast = input->Read(sizeof(start_timestamp), &start_timestamp);
+    assert(ast.ok());
```
Comment (cydrain) [inline internal/core/src/storage/Event.cpp:185]: need change all "assert" to "AssertInfo"

File: internal/core/src/storage/Types.h [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+#include "common/Types.h"
+
+namespace milvus::storage {
+
+using DataType = milvus::DataType;
+using Timestamp = milvus::Timestamp;
+
+const int32_t MAGIC_NUM = 0xfffabc;
+
+enum MediumType {
```
Comment (cydrain) [inline internal/core/src/storage/Types.h:28]: suggest to use "StorageType"

File: internal/core/src/storage/PayloadWriter.h [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    add_payload(const Payload& raw_data);
+
+    void
+    add_one_string_payload(const char* str, int str_size);
+
+    void
+    finish();
+
+    bool
+    has_finish();
```
Comment (cydrain) [inline internal/core/src/storage/PayloadWriter.h:42]: rename to "has_finished"

File: internal/core/src/storage/Event.cpp (hunk 10 above)
Comment (xige-16) [reply internal/core/src/storage/Event.cpp:41]: sizeof(header) may be larger than total sizeof(header.element)

File: internal/core/src/storage/DataCodec.h (hunk 6 above)
Comment (xige-16) [reply internal/core/src/storage/DataCodec.h:51]: We need the size  of the uint8 array

**Comment by xige-16** [comment]
rerun ut

**Comment by xige-16** [comment]
rerun ut

**Comment by xige-16** [comment]
rerun ut

**Comment by xige-16** [comment]
rerun ut

**Pull request merged**